Select map updates

gitriley opened the pull request.

vanwars commented:
Here were my manual testing procedures:

destroyed and recreate the database: 
```shell
$ vagrant ssh
$ DB_OWNER="postgres"
$ DB_NAME="lg_test_database"
$ HOST="localhost:7777"
$ psql -c "drop database $DB_NAME" -U $DB_OWNER
$ psql -c "create database $DB_NAME" -U $DB_OWNER
$ cd /localground/apps
$ python manage.py migrate # builds the database according to the branch specs
$ psql -c "update django_site set domain='$HOST', name='$HOST' where name = 'example.com'" -U $DB_OWNER -d $DB_NAME
```

vanwars commented on static/apps/style/views/left/layer-list-view.js:
```diff
@@ -33,56 +33,61 @@ define(["marionette",
                 if (this.app.layerHasBeenAltered && !this.layerHasBeenSaved) {
-                    console.log("should send save confirmation");
                     continueAction = confirm("You have unsaved changes on your currently selected layer. If you continue, your changes will not be saved. Do you wish to continue?");
                 }
```
Very nice confirm here.

vanwars commented on static/apps/style/style-app.js:
```diff
@@ -33,7 +33,7 @@ define([
             // kicks off any objects and processes that need to run
```
regarding the "layerHasBeenAltered" and "layerHasBeenSaved" flags: I think that there's a Backbone model property called model.hasChanged() that can be used to check (versus manually tracking this).

vanwars commented on static/apps/style/views/left/left-panel.js:
It may be more efficient / readable enforce the rule that every time a new map is selected, this LeftPanel view gets completely replaced (from the style-app file), and all four regions (PanelStylesView, LayerListView, SkinView, and SelectMapView) get initialized at once. To me, this conditional code that renders some things before the model and other things after the model is defined is confusing. It would be clearer (IMO) to enforce the rule that this view is always instantiated with a model. This may also solve the problem of things rendering more than once.

vanwars commented on static/apps/style/views/right/marker-style-view.js:
```diff
@@ -25,11 +25,7 @@ define(["jquery",
-                //'change:metadata': 'contData'
-            },
-
+            
```
It might make sense, down the line, to break this up into smaller helper views.  ~600 lines for a few is a bit long. That said, it's working well!

vanwars commented on static/tests/spec/views/style-left-layer-list-child-view-test.js:
```diff
@@ -80,10 +79,8 @@ define([
-                expect(layerListChildView.sendCollection).toHaveBeenCalledTimes(0);
-                fixture.find('.edit').trigger("click");
-                expect(layerListChildView.sendCollection).toHaveBeenCalledTimes(1);
+            it("router should trigger controller function", function() {
```
Please fix this last test: expect(1).toEqual(-1)

gitriley commented:
**Bug 1:** Fixed this bug by routing to the newly created map upon 'save'. However, the route ultimately happens twice due to the following chain of events:
'change-map' event —> handleNewMap() —> 'ready-for-routing' event —> rerouteIfNeeded()

This doesn't cause any problems, it's just weird. By removing the 'ready-for-routing' event, the routing only happens once, without problem. However, I'm sure we have that code in there to handle some other issue, so I'll leave it for now.

**Bug 2:** I was unable to recreate this, but I can try again using your exact process, or you can show me on Thursday/Friday.

**Bug 3:** Fixed.

- added pointer cursor to preview icon. I thought we were just going to get rid of the green preview button? I can add an href for the current map to the button, if that's what we want to do.

**Step 9 POINT OF CONFUSION: ** Fixed. Changes to the color ramp will no longer overwrite individual shape modifications. They will override individual color modifications, obviously.

**Step 11 Note** Not sure exactly what you mean, but we can talk about it. 

**NOTE: ** I added some code so that the app will reroute correctly after maps and layers are deleted. 
e.g. if the url is:
http://localhost:7777/style/#/277/layers/656

and the layer is deleted, it will reroute to:
http://localhost:7777/style/#/277

If the map is deleted, it reroutes to
http://localhost:7777/style/#

**Final Note: ** I've noticed that the loading time when switching between maps is extremely slow. Do you experience this? I'm going to go back a few commits and find out if it's been this way for awhile or if it's being caused by some new code.
